File: ios/RunWrap/OutfitRules.swift
import Foundation

/// 러닝 복장 아이템 — 룰(엔진)은 조합까지만 정하고 아이콘·라벨은 화면이 매핑한다
/// (RRTone과 같은 분리 — 엔진은 UI를 모른다)
enum OutfitItem: String, CaseIterable, Equatable {
    case singlet, shortSleeve, longSleeve, shorts, tights
    case jacket, gloves, windbreaker
    case waterproofCap, waterproofJacket
    case thermalTop, thermalBottom, beanie, neckWarmer
    case sunCap, sunglasses, sunscreen

    /// 노출 라벨 — 오늘 화면의 타일과 홈 판단 카드의 한 줄이 같은 표기를 쓰도록 한곳에 둔다.
    /// 아이콘 매핑은 여전히 화면 몫이다 (엔진은 UI를 모른다)
    var label: String {
        switch self {
        case .singlet: "싱글렛"
        case .shortSleeve: "반팔 티"
        case .longSleeve: "긴팔 티"
        case .shorts: "반바지"
        case .tights: "타이츠"
        case .jacket: "자켓"
        case .gloves: "장갑"
        case .windbreaker: "바람막이"
        case .waterproofCap: "방수 캡"
        case .waterproofJacket: "방수 자켓"
        case .thermalTop: "방한 상의"
        case .thermalBottom: "방한 하의"
        case .beanie: "비니"
        case .neckWarmer: "넥워머"
        case .sunCap: "러닝 캡"
        case .sunglasses: "선글라스"
        case .sunscreen: "선크림"
        }
    }
}

/// 복장 룰 — 체감온도 구간을 기본으로 습도·바람·강수·자외선·계절을 가산한다
/// (확장 요구, 2026-08-12). 구간은 계획서 M6 표, 가산 조건 출처:
/// - 자외선 보호 하한 UV 3: WHO Global Solar UV Index (Moderate부터 보호 권고)
/// - 고온다습·추위 레이어링: Nike 기온별 러닝 복장 가이드, 러닝 커뮤니티 겨울 복장 관례
/// 계절은 주입받은 now의 월로 판정한다 (기상학적 구분 — 여름 6~8월, 겨울 12~2월).
enum OutfitRules {
    /// 바람막이를 더하는 풍속 하한 (m/s)
    static let windbreakerMs = 8.0
    /// 자외선 보호 세트를 더하는 UV 지수 하한 (WHO Moderate)
    static let sunProtectionUV = 3.0

    static func outfit(apparentC: Double, humidityPct: Double, windMs: Double,
                       precipitationMm: Double, uvIndex: Double?, now: Date) -> [OutfitItem] {
        let raining = precipitationMm > 0
        var items: [OutfitItem]

        switch apparentC {
        case 24...:
            items = [.singlet, .shorts]
        case 16..<24:
            // 고온다습(≥80%)이면 땀이 증발하지 못해 한 단계 가볍게 — 반팔 대신 싱글렛
            items = humidityPct >= 80 ? [.singlet, .shorts] : [.shortSleeve, .shorts]
        case 8..<16:
            items = [.longSleeve, .tights]
        case 0..<8:
            items = [.longSleeve, .jacket, .tights, .gloves]
            // 겨울에는 같은 온도라도 귀 시림이 커 비니를 더한다
            if isWinter(now) { items.append(.beanie) }
        default:  // 영하
            items = [.thermalTop, .thermalBottom, .beanie, .neckWarmer, .gloves]
        }

        // 강수 가산 — 따뜻하면 챙으로 비만 막고(방수 캡), 서늘하면 체온 유지까지(방수 자켓)
        if raining {
            items.append(apparentC >= 16 ? .waterproofCap : .waterproofJacket)
        }

        // 바람 가산 — 8~24°C에서 바람막이. 24°C 이상 더위엔 겹옷이 역효과, 8°C 미만은 자켓이 겸한다
        if windMs >= windbreakerMs, (8..<24).contains(apparentC) {
            items.append(.windbreaker)
        }

        // 자외선 가산 — UV 3 이상이면 캡·선글라스·선크림 세트.
        // 여름에 UV 값이 없으면(야간 제외 API 누락) 보호 세트를 기본 포함한다.
        // 8°C 미만은 비니 영역이라 제외, 우천 시에는 이미 해가 없어 제외.
        let uv = uvIndex ?? (isSummer(now) ? sunProtectionUV : 0)
        if uv >= sunProtectionUV, apparentC >= 8, !raining {
            items.append(contentsOf: [.sunCap, .sunglasses, .sunscreen])
        }

        return items
    }

    private static func month(_ now: Date) -> Int {
        Calendar.current.component(.month, from: now)
    }

    private static func isSummer(_ now: Date) -> Bool { (6...8).contains(month(now)) }
    private static func isWinter(_ now: Date) -> Bool { month(now) == 12 || month(now) <= 2 }
}

File: android/app/src/main/java/com/jkpark/runwrap/engine/OutfitRules.kt
package com.jkpark.runwrap.engine

import java.time.Instant

/// 러닝 복장 아이템 — 노출 라벨과 저장값 분리 (iOS rawValue 대응)
enum class OutfitItem(val storageValue: String) {
    SINGLET("singlet"), SHORT_SLEEVE("shortSleeve"), LONG_SLEEVE("longSleeve"),
    SHORTS("shorts"), TIGHTS("tights"), JACKET("jacket"), GLOVES("gloves"),
    WINDBREAKER("windbreaker"), WATERPROOF_CAP("waterproofCap"), WATERPROOF_JACKET("waterproofJacket"),
    THERMAL_TOP("thermalTop"), THERMAL_BOTTOM("thermalBottom"), BEANIE("beanie"),
    NECK_WARMER("neckWarmer"), SUN_CAP("sunCap"), SUNGLASSES("sunglasses"), SUNSCREEN("sunscreen");

    val label: String
        get() = when (this) {
            SINGLET -> "싱글렛"
            SHORT_SLEEVE -> "반팔 티"
            LONG_SLEEVE -> "긴팔 티"
            SHORTS -> "반바지"
            TIGHTS -> "타이츠"
            JACKET -> "자켓"
            GLOVES -> "장갑"
            WINDBREAKER -> "바람막이"
            WATERPROOF_CAP -> "방수 캡"
            WATERPROOF_JACKET -> "방수 자켓"
            THERMAL_TOP -> "방한 상의"
            THERMAL_BOTTOM -> "방한 하의"
            BEANIE -> "비니"
            NECK_WARMER -> "넥워머"
            SUN_CAP -> "러닝 캡"
            SUNGLASSES -> "선글라스"
            SUNSCREEN -> "선크림"
        }
}

/// 복장 추천 규칙 — 체감온도·습도·바람·비·자외선으로 러닝 복장을 조합한다.
/// 러닝 커뮤니티의 "달릴 때는 체감보다 10도 더운 옷" 관행을 체감온도 구간표로 옮긴 것.
/// iOS `OutfitRules.swift` 이식.
object OutfitRules {
    private const val WINDBREAKER_MS = 8.0     // 이 이상 바람이면 바람막이 (간절기 한정)
    private const val SUN_PROTECTION_UV = 3.0  // WHO "보통" 구간 시작 — 이 이상이면 햇빛 대비

    /// 추천 복장 목록 — 기본 복장(체감온도 구간) + 비·바람·자외선 조건 추가.
    /// uvIndex가 null이면 계절로 추정한다 (여름 한낮 러닝이 흔하므로 여름엔 보통(3.0) 취급)
    fun outfit(
        apparentC: Double,
        humidityPct: Double,
        windMs: Double,
        precipitationMm: Double,
        uvIndex: Double?,
        now: Instant,
    ): List<OutfitItem> {
        val raining = precipitationMm > 0
        val month = now.atZone(SEOUL).monthValue
        val isSummer = month in 6..8
        val isWinter = month == 12 || month == 1 || month == 2

        val items = mutableListOf<OutfitItem>()
        when {
            apparentC >= 24 -> items += listOf(OutfitItem.SINGLET, OutfitItem.SHORTS)
            apparentC >= 16 ->
                // 후텁지근하면(습도 80% 이상) 한 단계 가볍게
                items += if (humidityPct >= 80) {
                    listOf(OutfitItem.SINGLET, OutfitItem.SHORTS)
                } else {
                    listOf(OutfitItem.SHORT_SLEEVE, OutfitItem.SHORTS)
                }
            apparentC >= 8 -> items += listOf(OutfitItem.LONG_SLEEVE, OutfitItem.TIGHTS)
            apparentC >= 0 -> {
                items += listOf(OutfitItem.LONG_SLEEVE, OutfitItem.JACKET, OutfitItem.TIGHTS, OutfitItem.GLOVES)
                if (isWinter) items += OutfitItem.BEANIE
            }
            else -> items += listOf(
                OutfitItem.THERMAL_TOP, OutfitItem.THERMAL_BOTTOM,
                OutfitItem.BEANIE, OutfitItem.NECK_WARMER, OutfitItem.GLOVES,
            )
        }

        if (raining) {
            // 더울 땐 비를 맞고 뛰는 게 낫다 — 시야 확보용 캡만. 쌀쌀하면 저체온 대비 자켓
            items += if (apparentC >= 16) OutfitItem.WATERPROOF_CAP else OutfitItem.WATERPROOF_JACKET
        }
        if (windMs >= WINDBREAKER_MS && apparentC >= 8 && apparentC < 24) {
            items += OutfitItem.WINDBREAKER
        }
        val uv = uvIndex ?: if (isSummer) 3.0 else 0.0
        if (uv >= SUN_PROTECTION_UV && apparentC >= 8 && !raining) {
            items += listOf(OutfitItem.SUN_CAP, OutfitItem.SUNGLASSES, OutfitItem.SUNSCREEN)
        }
        return items
    }
}
